### PR TITLE
nixos/kavita: init

### PR DESCRIPTION
**Pull request by Misterio77**
###### Description of changes

Add a module for https://github.com/NixOS/nixpkgs/pull/227890

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/web-apps/kavita.nix [hunk 1]
```diff
@@ -0,0 +1,84 @@
+{ config, lib, pkgs, ... }:
+
+let cfg = config.services.kavita;
+in {
```
Comment (AndersonTorres) [inline nixos/modules/services/web-apps/kavita.nix:4]:
```suggestion
let 
  cfg = config.services.kavita;
in {
```

File: nixos/modules/services/web-apps/kavita.nix [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+      default = "/var/lib/kavita";
+      type = lib.types.str;
+      description = lib.mdDoc "The directory where Kavita stores its state.";
+    };
+
+    tokenKeyFile = lib.mkOption {
+      type = lib.types.path;
+      description = lib.mdDoc ''
+        A file containing the TokenKey, a secret with at 128+ bits. One can be
+        generated with `head -c 32 /dev/urandom | base64`.
```
Comment (AndersonTorres) [inline nixos/modules/services/web-apps/kavita.nix:31]:
```suggestion
        A file containing the TokenKey, a secret with at 128+ bits. 
        It can be generated with `head -c 32 /dev/urandom | base64`.
```

File: nixos/tests/all-tests.nix [hunk 3]
```diff
@@ -341,6 +341,7 @@ in {
   kafka = handleTest ./kafka.nix {};
   kanidm = handleTest ./kanidm.nix {};
   karma = handleTest ./karma.nix {};
+  kavita = handleTest ./kavita.nix {};
```
Comment (AndersonTorres) [inline nixos/tests/all-tests.nix:344]: Split tests in a separate commit.

File: pkgs/servers/web-apps/kavita/default.nix [hunk 4]
```diff
@@ -0,0 +1,63 @@
+{
+  lib,
+  fetchFromGitHub,
+  buildDotnetModule,
+  dotnetCorePackages,
+  buildNpmPackage,
+  nixosTests,
+}:
```
Comment (AndersonTorres) [inline pkgs/servers/web-apps/kavita/default.nix:8]:
```suggestion
{ lib
, fetchFromGitHub
, buildDotnetModule
, buildNpmPackage
, dotnetCorePackages
, nixosTests
}:

```

File: pkgs/servers/web-apps/kavita/default.nix [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+buildDotnetModule rec {
+  pname = "kavita";
+  version = "0.7.1.4";
+
+  src = fetchFromGitHub {
+    owner = "kareadita";
+    repo = pname;
```
Comment (AndersonTorres) [inline pkgs/servers/web-apps/kavita/default.nix:15]:
```suggestion
    repo = "kavita";
```

File: pkgs/servers/web-apps/kavita/default.nix [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  nugetDeps = ./nuget-deps.nix;
+  dotnet-sdk = dotnetCorePackages.sdk_6_0;
+  dotnet-runtime = dotnetCorePackages.aspnetcore_6_0;
+
+  passthru.tests = { inherit (nixosTests) kavita; };
+
+  meta = {
+    description = "A fast, feature rich, cross platform reading server";
+    homepage = "https://kavitareader.com";
+    changelog = "https://github.com/${src.owner}/${src.repo}/releases/tag/${src.rev}";
```
Comment (AndersonTorres) [inline pkgs/servers/web-apps/kavita/default.nix:58]:
```suggestion
    changelog = "https://github.com/kareadita/kavita/releases/tag/${src.rev}";
```

No need to parameterize everything.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
The commit for the module should be 

```
nixos/kavita: init
```

for tests:

```
nixos/tests/kavita: init
```

**Comment by Misterio77** [comment]
Thanks! Feedback applied.

**Review by natsukium — APPROVED** [review]

File: pkgs/servers/web-apps/kavita/default.nix [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    sourceRoot = "source/UI/Web";
+
+    npmBuildScript = "prod";
+    npmFlags = [ "--legacy-peer-deps" ];
+    npmRebuildFlags = [ "--ignore-scripts" ]; # Prevent playwright from trying to install browsers
+    npmDepsHash = "sha256-w0CuTPyCQyAxULvqd6+GiZaPlO8fh4xLmbEnGA47pL8=";
+  };
+
+  postPatch = ''
+    sed -i "s:@WEB_ROOT@:$out/lib/kavita-frontend:" API/Controllers/FallbackController.cs API/Startup.cs
```
Comment (AndersonTorres) [inline pkgs/servers/web-apps/kavita/default.nix:38]: There is a specific function for those @@-substitutions, namely `substituteAll`. 
Look at the example:

https://github.com/NixOS/nixpkgs/blob/6027cc8c1c4567d8e6d02c5ac1e4b8ff1193c08d/pkgs/tools/system/wsysmon/default.nix#L26-L29

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/servers/web-apps/kavita/default.nix (hunk 7 above)
Comment (Misterio77) [reply pkgs/servers/web-apps/kavita/default.nix:38]: Wow that's pretty nice. There really is a function for everything

Comment (Misterio77) [reply pkgs/servers/web-apps/kavita/default.nix:38]: There seems to be an issue with this approach.

It apparently can't access `$out` or `$frontend`, so the only way is to use `${frontend}`, but then the frontend attribute won't be overridable (which is something that seems worth keeping). Any ideas?

Comment (Misterio77) [reply pkgs/servers/web-apps/kavita/default.nix:38]: Oh apparently I can use the `finalAttrs` approach instead.

Comment (Misterio77) [reply pkgs/servers/web-apps/kavita/default.nix]: The solution I came up with is using mkDerivation with `backend` and `frontend` as attrs, let me know what you think.

**Review by AndersonTorres — APPROVED** [review]

File: pkgs/servers/web-apps/kavita/default.nix (hunk 7 above)
Comment (AndersonTorres) [reply pkgs/servers/web-apps/kavita/default.nix]: Unfortunately not all `build<ecosystem>Package`ers were converted to accept the `(finalAttrs: {...})` construction.

File: nixos/modules/services/web-apps/kavita.nix [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    systemd.services.kavita = {
+      description = "Kavita";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
+      preStart = ''
+        mkdir -p "${cfg.dataDir}/config"
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/kavita.nix:54]: We normally use tmpfiles.d to create directories for services

File: nixos/modules/services/web-apps/kavita.nix [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        cat > "${cfg.dataDir}/config/appsettings.json" <<EOF
+        {
+          "TokenKey": "$(cat ${cfg.tokenKeyFile})",
+          "Port": ${toString cfg.port},
+          "IpAddresses": "${lib.concatStringsSep "," cfg.ipAdresses}"
+        }
+        EOF
+        chmod 640 ${cfg.dataDir}/config/appsettings.json
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/kavita.nix:62]: Umask should be set so that this is not required, otherwise there is a short window where the file has probably more open permissions

File: nixos/modules/services/web-apps/kavita.nix [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+      };
+    };
+
+    users = {
+      users.${cfg.user} = {
+        description = "kavita service user";
+        isSystemUser = true;
+        group = cfg.user;
+        home = cfg.dataDir;
+        createHome = true;
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/kavita.nix:78]: Often systemusers do not have their home directory created because it contains skel files which such a user does not need

File: nixos/tests/kavita.nix [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+      password = "correcthorsebatterystaple";
+      email = "foo@bar";
+    });
+    manifestUrl = "http://kavita:5000/site.webmanifest";
+  in ''
+    kavita.start
+    kavita.wait_for_unit("kavita.service")
+
+    # Check that static assets are working
+    kavita.wait_until_succeeds("curl ${manifestUrl} | grep Kavita")
```
Comment (SuperSandro2000) [inline nixos/tests/kavita.nix:30]:
```suggestion
  in ''
    kavita.start
    kavita.wait_for_unit("kavita.service")

    # Check that static assets are working
    kavita.wait_until_succeeds("curl "http://kavita:5000/site.webmanifest" | grep Kavita")
```

File: nixos/modules/services/web-apps/kavita.nix (hunk 10 above)
Comment (Misterio77) [reply nixos/modules/services/web-apps/kavita.nix:78]: I see. I wonder if there's there a recommended way to create these data directories?

Comment (Misterio77) [reply nixos/modules/services/web-apps/kavita.nix:78]: It seems that tmpfiles is the most common way (even for persistent stuff), I'll do it that way, then.

**Comment by Misterio77** [comment]
Thanks for the review, @SuperSandro2000! Addressed all of your comments, let me know if you have any other suggestions!

**Comment by Misterio77** [comment]
I've used `tmpfiles` to create the `dataDir` and `${dataDir}/config`. I set them to be `750`, does that look good? Or should I use `700` instead?

File: pkgs/servers/web-apps/kavita/default.nix [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+  };
+
+  dontBuild = true;
+
+  installPhase = ''
+    mkdir -p $out/bin $out/lib/kavita
+    ln -s $backend/lib/kavita-backend $out/lib/kavita/backend
+    ln -s $frontend/lib/node_modules/kavita-webui/dist $out/lib/kavita/frontend
+    ln -s $backend/bin/API $out/bin/kavita
+  '';
```
Comment (natsukium) [inline pkgs/servers/web-apps/kavita/default.nix:61]:
```suggestion
  installPhase = ''
    runHook preInstall
  
    mkdir -p $out/bin $out/lib/kavita
    ln -s $backend/lib/kavita-backend $out/lib/kavita/backend
    ln -s $frontend/lib/node_modules/kavita-webui/dist $out/lib/kavita/frontend
    ln -s $backend/bin/API $out/bin/kavita
    
    runHook postInstall
  '';
```

Comment (Misterio77) [reply pkgs/servers/web-apps/kavita/default.nix]: Thanks!

**Review by thiagokokada — APPROVED** [review]
